File: server/modules/hardware/services/catalog.ts
/**
 * Migrado de legacy/backend/modules/batteries/batteries.catalog.ts.
 */

/**
 * Catálogo canónico único: toda bateria do sistema é `battery_estelar`, infinita
 * por design. Migração `20260516120000_all_batteries_become_estelar` colapsa todas
 * as instâncias e o stock para este id; o normalizador abaixo cobre referências
 * em código/payloads enviados por clientes antigos.
 */
export const CANONICAL_1000WH_BATTERY_ID = 'battery_estelar';

/**
 * IDs removidos do catálogo em `20260516200000_purge_all_legacy_battery_and_charger_residue`.
 * Stock/save-game não devem reintroduzi-los (evita placeholders `temp_legacy_*`).
 * Nota: `battery_nebula`/`nebula` NÃO entram aqui — Nébula Free-to-Play é catálogo real.
 */
export const PURGED_LEGACY_STOCK_IDS = new Set(['charger_a1', 'charger_a2', 'battery_aa', 'battery_car', 'battery_diesel', 'battery_fusion', 'battery_protostar', 'battery_ups', 'battery_wall', 'small_battery', 'supernova']);

/** Subconjunto expurgado que deve fundir-se em `battery_estelar` (não carregadores). */
export const PURGED_LEGACY_STOCK_REMAP_TO_ESTELAR = new Set([...PURGED_LEGACY_STOCK_IDS].filter((id) => !id.startsWith('charger_')));

/** Ids legados conhecidos que devem ser tratados como `battery_estelar`. */
export const LEGACY_1000WH_BATTERY_IDS = new Set(['small_battery', 'battery_protostar', 'battery_stellar']);

export const KNOWN_INFINITE_BATTERY_IDS = new Set(['battery_estelar', 'battery_protostar', 'battery_stellar']);

/** Remapeia ids legados de bateria 1000Wh (`small_battery`, `battery_protostar`, `battery_stellar`) para o catálogo canónico `battery_estelar`. */
export function normalizeKnown1000WhBatteryCatalogId(itemIdRaw: unknown): string {
  const itemId = itemIdRaw == null ? '' : String(itemIdRaw).trim();
  if (!itemId) return '';
  return LEGACY_1000WH_BATTERY_IDS.has(itemId) ? CANONICAL_1000WH_BATTERY_ID : itemId;
}

/** Ids de itens expurgados do catálogo: baterias antigas viram `battery_estelar`, carregadores viram `''` (descartados). */
export function remapPurgedStockItemId(itemIdRaw: unknown): string {
  const itemId = itemIdRaw == null ? '' : String(itemIdRaw).trim();
  if (!itemId) return '';
  if (PURGED_LEGACY_STOCK_REMAP_TO_ESTELAR.has(itemId)) return CANONICAL_1000WH_BATTERY_ID;
  if (PURGED_LEGACY_STOCK_IDS.has(itemId)) return '';
  return itemId;
}

/** Normalização de chaves de `stock` no save-game e saneamento de BD. */
export function normalizeStockCatalogItemId(itemIdRaw: unknown): string {
  const legacy = normalizeKnown1000WhBatteryCatalogId(itemIdRaw);
  return remapPurgedStockItemId(legacy);
}

/**
 * `placed_racks.battery_id` pode ser o id de catálogo (upgrades.id) ou o id de instância em `stored_batteries`.
 * Cron/ranking devem resolver para o id de catálogo antes de consultar `upgrades`.
 */
export function resolvePlacedRackBatteryCatalogId(
  batteryIdRaw: unknown,
  storedInstanceIdToCatalogId: Map<string, string>,
  snapshotCatalogIdRaw?: unknown
): string {
  const s = batteryIdRaw == null ? '' : String(batteryIdRaw).trim();
  if (!s) return '';
  const snap = snapshotCatalogIdRaw == null ? '' : normalizeKnown1000WhBatteryCatalogId(snapshotCatalogIdRaw);
  if (snap) return snap;
  return storedInstanceIdToCatalogId.get(s) || s;
}

const LEGACY_TEMP_ID_MAX_LEN = 200;
const LEGACY_TEMP_SLUG_MAX_LEN = 80;

/** ID determinístico para placeholder legacy (sem contador global por arranque). */
/** Id determinístico (`temp_legacy_{userId}_{slug}`) para placeholder de item legado sem catálogo — mesmo input sempre gera o mesmo id. */
export function buildStableLegacyTempUpgradeId(userId: number, normalizedOriginal: string): string {
  const slug =
    normalizedOriginal
      .toLowerCase()
      .replace(/[^a-z0-9_.-]+/g, '-')
      .replace(/^-+|-+$/g, '')
      .slice(0, LEGACY_TEMP_SLUG_MAX_LEN) || 'sem-id';
  return `temp_legacy_${userId}_${slug}`.slice(0, LEGACY_TEMP_ID_MAX_LEN);
}

/**
 * Sistema de baterias é infinito por design: qualquer bateria existente é tratada
 * como ilimitada, sem necessidade de carregar. Mantemos guard para id vazio para
 * preservar semântica "sem bateria equipada" nos chamadores.
 */
export function isKnownInfiniteBatteryCatalogId(itemIdRaw: unknown): boolean {
  const itemId = normalizeKnown1000WhBatteryCatalogId(itemIdRaw).toLowerCase();
  return itemId.length > 0;
}

File: rust/genesis-core/src/hardware/catalog.rs
//! Hardware catalog id normalization — 1:1 with `server/modules/hardware/services/catalog.ts`.

pub const CANONICAL_1000WH_BATTERY_ID: &str = "battery_estelar";

/// IDs purged in `20260516200000_purge_all_legacy_battery_and_charger_residue`.
pub const PURGED_LEGACY_STOCK_IDS: &[&str] = &[
    "charger_a1",
    "charger_a2",
    "battery_aa",
    "battery_car",
    "battery_diesel",
    "battery_fusion",
    "battery_protostar",
    "battery_ups",
    "battery_wall",
    "small_battery",
    "supernova",
];

pub const LEGACY_1000WH_BATTERY_IDS: &[&str] =
    &["small_battery", "battery_protostar", "battery_stellar"];

pub const KNOWN_INFINITE_BATTERY_IDS: &[&str] =
    &["battery_estelar", "battery_protostar", "battery_stellar"];

const CHARGER_PREFIX: &str = "charger_";

fn trim_item_id(item_id_raw: Option<&str>) -> String {
    item_id_raw.unwrap_or("").trim().to_string()
}

fn is_purged_legacy_stock_id(item_id: &str) -> bool {
    PURGED_LEGACY_STOCK_IDS.iter().any(|id| *id == item_id)
}

fn is_purged_remap_to_estelar(item_id: &str) -> bool {
    is_purged_legacy_stock_id(item_id) && !item_id.starts_with(CHARGER_PREFIX)
}

/// Remaps known 1000Wh legacy ids (`small_battery`, `battery_protostar`, `battery_stellar`)
/// to the canonical `battery_estelar`.
pub fn normalize_known_1000wh_battery_catalog_id(item_id_raw: Option<&str>) -> String {
    let item_id = trim_item_id(item_id_raw);
    if item_id.is_empty() {
        return String::new();
    }
    if LEGACY_1000WH_BATTERY_IDS.iter().any(|id| *id == item_id) {
        CANONICAL_1000WH_BATTERY_ID.to_string()
    } else {
        item_id
    }
}

/// Purged batteries remap to Estelar; purged chargers become `""` (dropped).
pub fn remap_purged_stock_item_id(item_id_raw: Option<&str>) -> String {
    let item_id = trim_item_id(item_id_raw);
    if item_id.is_empty() {
        return String::new();
    }
    if is_purged_remap_to_estelar(&item_id) {
        return CANONICAL_1000WH_BATTERY_ID.to_string();
    }
    if is_purged_legacy_stock_id(&item_id) {
        return String::new();
    }
    item_id
}

/// Save-game / DB stock key normalization.
pub fn normalize_stock_catalog_item_id(item_id_raw: Option<&str>) -> String {
    let legacy = normalize_known_1000wh_battery_catalog_id(item_id_raw);
    remap_purged_stock_item_id(Some(&legacy))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_legacy_1000wh_to_estelar() {
        assert_eq!(
            normalize_known_1000wh_battery_catalog_id(Some("small_battery")),
            CANONICAL_1000WH_BATTERY_ID
        );
        assert_eq!(
            normalize_known_1000wh_battery_catalog_id(Some("battery_stellar")),
            CANONICAL_1000WH_BATTERY_ID
        );
        assert_eq!(
            normalize_known_1000wh_battery_catalog_id(Some("battery_protostar")),
            CANONICAL_1000WH_BATTERY_ID
        );
    }

    #[test]
    fn normalize_keeps_non_legacy() {
        assert_eq!(
            normalize_known_1000wh_battery_catalog_id(Some("battery_nebula")),
            "battery_nebula"
        );
        assert_eq!(normalize_known_1000wh_battery_catalog_id(None), "");
        assert_eq!(normalize_known_1000wh_battery_catalog_id(Some("  ")), "");
    }

    #[test]
    fn remap_purged_battery_to_estelar_charger_dropped() {
        assert_eq!(
            remap_purged_stock_item_id(Some("battery_aa")),
            CANONICAL_1000WH_BATTERY_ID
        );
        assert_eq!(remap_purged_stock_item_id(Some("charger_a1")), "");
        assert_eq!(remap_purged_stock_item_id(Some("gpu_rx6600")), "gpu_rx6600");
    }

    #[test]
    fn normalize_stock_catalog_collapses_legacy_then_purge() {
        assert_eq!(
            normalize_stock_catalog_item_id(Some("small_battery")),
            CANONICAL_1000WH_BATTERY_ID
        );
        assert_eq!(normalize_stock_catalog_item_id(Some("charger_a2")), "");
        assert_eq!(
            normalize_stock_catalog_item_id(Some("battery_estelar")),
            CANONICAL_1000WH_BATTERY_ID
        );
    }
}
